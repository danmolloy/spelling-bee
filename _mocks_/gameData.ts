import { faker } from "@faker-js/faker"

//const centerLetter = faker.string.alpha();
//const outerLetters = faker.string.alpha(6).split("")
const expiration = Date.now() + 10000
const answers = faker.word.words(25).toLowerCase().split("").filter(i => i !== " ")
const lettersArr = Array.from(new Set(answers)).slice(0, 7)

const validLetters = Array.from(new Set(faker.string.alpha(12).split(""))).slice(0, 7)

export const mockData = {
  expiration: expiration,
  displayWeekday:"Sunday",
  displayDate: "October 13, 1986",
  printDate:"2022-04-10",
  centerLetter: lettersArr[0],
  outerLetters: lettersArr.slice(1, 7),
  validLetters: lettersArr,
  pangrams: [lettersArr.join("")],
  answers: [lettersArr.join(""), lettersArr.slice(1, 6).join(""), ...faker.word.words(25).split(" ")],
  id: faker.string.uuid(),
  freeExpiration:"0",
  editor:"Sam Ezersky"
}