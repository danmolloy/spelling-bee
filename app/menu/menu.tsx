import Link from "next/link";
import { AiOutlineClose } from "react-icons/ai";

export const menuItems = [
  {
    name: "How to Play",
    onClickArg: "howTo",
    href: "how-to"
  },
  {
    name: "Rankings",
    onClickArg: "rankings",
    href: "how-to"

  },
  {
    name: "Today's Hints",
    onClickArg: "hints",
    href: "hints"
  },
]

export type MenuProps = {
  setShowMenuItem: (arg: string|null) => void
}

export default function Menu(props: MenuProps) {
  const { setShowMenuItem  } = props;
  
  return (
  <div data-testid="menu-page" className="h-screen z-40 shadow w-full backdrop-blur-sm absolute">
      <div className="mx-2 rounded mt-12 p-2 md:mx-24 lg:mx-60 shadow-md border flex flex-col  bg-white text-zinc-700 dark:bg-zinc-900 dark:text-zinc-300 dark:border dark:border-zinc-800">
      <div data-testid="nav-menu" className=" flex flex-row justify-end dark:text-zinc-400 ">
        <button className="text-lg dark:hover:text-zinc-300 hover:text-yellow-500 p-1 rounded-full" onClick={() => setShowMenuItem(null)} data-testid="close-btn">
          <AiOutlineClose />
        </button>
      </div>
      {menuItems.map(i => (
        <button key={i.name} className="font-light border-b dark:border-zinc-800 py-2 my-1 px-2 hover:text-yellow-500" onClick={() => setShowMenuItem(i.onClickArg)}>
          <p>{i.name}</p>
        </button>
      ))}
    </div>
    </div>
  )
}