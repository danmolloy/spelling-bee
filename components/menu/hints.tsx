import { AiOutlineClose } from 'react-icons/ai'
import MenuPage from './menuPage'
import HelpTable from './helpTable'
import { getPoints } from '../../lib/functions'
import TwoLetterList from './twoLetterList'

export type HintsProps = {
  setShowMenuItem: (arg: string|null) => void
  answers: string[]
  pangrams: string[]
  setRevealAnswers: () => void,
  revealAnswers: boolean
  centerLetter: string
  validLetters: string[]
}

export default function Hints(props: HintsProps) {
  const { centerLetter, validLetters, setShowMenuItem, answers, pangrams, setRevealAnswers } = props;

  const reveal = () => {
    setShowMenuItem(null)
    setRevealAnswers();
    localStorage.setItem('revealed', "true");
  }
  
  return (
    <MenuPage>
      <div data-testid="hints-div" >
      <div className='flex flex-row w-full justify-between items-center'>
        <h1 className='text-2xl font-bold'>Hints</h1>
        <button data-testid="menu-icon" className='menu-icon' onClick={() => setShowMenuItem(null)}>
          <AiOutlineClose />
        </button>
      </div>
      <div className='p-2'>
        <p className='text-lg font-thin'>Words: {answers.length}</p>
        <p className='text-lg font-thin'>Pangrams: {pangrams.length}</p>
        <p className='text-lg font-thin'>Points: {getPoints(answers)}</p>
      </div>
      <div className='flex flex-col items-center py-2'>
        <HelpTable validLetters={validLetters} centerLetter={centerLetter} answers={answers} />
      </div>
      <TwoLetterList answers={answers} />
      <hr />
      <h2 className='text-xl font-medium py-1 my-2'>Reveal Answers</h2>
      <div className='how-to-section flex flex-col'>
        <p className='font-semithin'>Reveal the answer list for the current game.</p>
        <button className='self-center single-btn border-0 m-1 text-bold bg-yellow-300 hover:bg-yellow-200 active:bg-black active:text-yellow-300' onClick={() => reveal()}>Reveal</button>
      </div> 
      </div>
    </MenuPage>
  )
}