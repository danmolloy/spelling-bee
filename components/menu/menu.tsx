import { AiOutlineClose } from "react-icons/ai";
import MenuPage from "./menuPage";

export const menuItems = [
  {
    name: "How to Play",
    onClickArg: "howTo"
  },
  {
    name: "Rankings",
    onClickArg: "rankings"
  },
  {
    name: "Today's Hints",
    onClickArg: "hints"
  },
]

export type MenuProps = {
  setShowMenuItem: (arg: string) => void
}

export default function Menu(props: MenuProps) {
  const { setShowMenuItem  } = props;
  
  return (
    <MenuPage>
      <div data-testid="nav-menu" className="mb-2 flex flex-row justify-end dark:text-zinc-400 ">
        <button className="text-lg dark:hover:text-zinc-300" onClick={() => setShowMenuItem(null)} data-testid="close-btn">
          <AiOutlineClose />
        </button>
      </div>
      {menuItems.map(i => (
        <button key={i.name} className="font-light border-b dark:border-zinc-800 py-2 my-1 px-2" onClick={() => setShowMenuItem(i.onClickArg)}>
          <p>{i.name}</p>
        </button>
      ))}
    </MenuPage>
  )
}