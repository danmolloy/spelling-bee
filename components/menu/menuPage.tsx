
export type MenuPageProps = {
  children: React.ReactNode
}

export default function MenuPage(props: MenuPageProps) {
  const { children } = props;

  return (
    <div data-testid="menu-page" className="mt-[7vh] h-[93vh] md:mt-[14vh] md:h-[86vh] z-40 shadow w-full backdrop-blur-sm absolute">
      <div className="mt-8 p-2 md:mx-24 lg:mx-60 shadow-md border flex flex-col  bg-white text-zinc-700 dark:bg-zinc-900 dark:text-zinc-300 dark:border dark:border-zinc-800">
        {children}
      </div>
    </div>
  );
}