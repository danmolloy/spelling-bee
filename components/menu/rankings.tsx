import { useState, useEffect } from 'react'
import { AiOutlineClose } from 'react-icons/ai'
import { getPoints } from '../../lib/functions'
import MenuPage from './menuPage'

export const rankingLevels = [
  {
    keyIndex: 0,
    name: "Beginner",
    minScoreMultiplier: 0
  },
  {
    keyIndex: 1,
    name: "Good Start",
    minScoreMultiplier: .125
  },
  {
    keyIndex: 2,
    name: "Moving Up", 
    minScoreMultiplier: .25
  },
  {
    keyIndex: 3,
    name: "Good",
    minScoreMultiplier:  0.375
  },
  {
    keyIndex: 4,
    name: "Solid",
    minScoreMultiplier: 0.5
  },
  {
    keyIndex: 5,
    name: "Nice",
    minScoreMultiplier: 0.625
  },
  {
    keyIndex: 6,
    name: "Great",
    minScoreMultiplier: 0.74
  },
  {
    keyIndex: 7,
    name: "Amazing",
    minScoreMultiplier: 0.865
  },
  {
    keyIndex: 8,
    name: "Genius",
    minScoreMultiplier: 1
  },
]

export type RankingsProps = {
  answers: string[]
  setShowMenuItem: (arg: null) => void
}


export default function Rankings(props: RankingsProps) {
  const { answers, setShowMenuItem } = props;
  const [answersSum, setAnswersSum] = useState(null)

  useEffect(() => {
    if (!answers) {
      return
    }
    setAnswersSum(getPoints(answers));
  }, [answers])

  return (
      <MenuPage>
      <div data-testid="rankings-div" className='flex flex-row justify-between'>
        <h2 className="font-bold text-2xl">Rankings</h2>
        <button data-testid="menu-icon" className='menu-icon' onClick={() => setShowMenuItem(null)}>
          <AiOutlineClose />
        </button>
      </div>
      <div data-testid="help-text" className='p-4 font-semithin text-lg'>   
        <p>Below are the minimum scores for the rankings of the current puzzle.</p>
        <p>Please refer to <i>How to Play</i> for scoring system.</p>
      </div>
      <ul className="pl-4 font-semithin">
        {rankingLevels.map(i => (
          <li data-testid={`${i.name}-level`} key={i.name}>{i.name} (<span className='font-medium'>{Math.floor(answersSum *i.minScoreMultiplier)}</span>)</li>
        ))}
      </ul>
    </MenuPage>
    
  )
}