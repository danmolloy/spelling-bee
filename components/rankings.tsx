import { useState, useEffect } from 'react'
import { AiOutlineClose } from 'react-icons/ai'
import { answerSum } from '../pages/index'

export const rankingLevels = [
  {
    name: "Beginner",
    minScoreMultiplier: 0
  },
  {
    name: "Good Start",
    minScoreMultiplier: .125
  },
  {
    name: "Moving Up", 
    minScoreMultiplier: .25
  },
  {
    name: "Good",
    minScoreMultiplier:  0.375
  },
  {
    name: "Solid",
    minScoreMultiplier: 0.5
  },
  {
    name: "Nice",
    minScoreMultiplier: 0.625
  },
  {
    name: "Great",
    minScoreMultiplier: 0.74
  },
  {
    name: "Amazing",
    minScoreMultiplier: 0.865
  },
  {
    name: "Genius",
    minScoreMultiplier: 1
  },
]


export default function Rankings({ data, showRankingsToggle }) {
  const [answersSum, setAnswersSum] = useState(null)

  useEffect(() => {
    if (!data) {
      return
    }
    setAnswersSum(answerSum(data?.gameData.yesterday.answers, data?.gameData.yesterday.pangrams));
  }, [data])

  return (
      <div className="slide-up rankings" onBlur={() => showRankingsToggle()}>
      <div className='flex flex-row justify-between'>
        <h2 className="font-bold text-2xl">Rankings</h2>
        <button className='menu-icon' onClick={() => showRankingsToggle()}>
          <AiOutlineClose />
        </button>
      </div>
      <div className='p-4 font-semithin text-lg'>   
        <p >Below are the minimum scores for the rankings of the current puzzle.</p>
        <p> Please refer to <i>How to Play</i> for scoring system.</p>
      </div>
      <ul className="pl-4 font-semithin">
        {rankingLevels.map(i => (
          <li key={i.name}>{i.name} (<span className='font-medium'>{Math.floor(answersSum *i.minScoreMultiplier)}</span>)</li>
        ))}
      </ul>
    </div>
    
  )
}