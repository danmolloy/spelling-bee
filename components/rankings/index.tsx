import { useEffect, useRef } from 'react'
import { AiOutlineClose } from 'react-icons/ai'

export const rankingLevels = [
  {
    keyIndex: 0,
    name: "Beginner",
    minScoreMultiplier: 0
  },
  {
    keyIndex: 1,
    name: "Good Start",
    minScoreMultiplier: .125
  },
  {
    keyIndex: 2,
    name: "Moving Up", 
    minScoreMultiplier: .25
  },
  {
    keyIndex: 3,
    name: "Good",
    minScoreMultiplier:  0.375
  },
  {
    keyIndex: 4,
    name: "Solid",
    minScoreMultiplier: 0.5
  },
  {
    keyIndex: 5,
    name: "Nice",
    minScoreMultiplier: 0.625
  },
  {
    keyIndex: 6,
    name: "Great",
    minScoreMultiplier: 0.74
  },
  {
    keyIndex: 7,
    name: "Amazing",
    minScoreMultiplier: 0.865
  },
  {
    keyIndex: 8,
    name: "Genius",
    minScoreMultiplier: 1
  },
]

export type RankingsProps = {
  setShowMenuItem: (arg: null) => void
  geniusScore: number
  currentScore: number
}


export default function Rankings(props: RankingsProps) {
  const { setShowMenuItem, geniusScore, currentScore } = props;
  const ref = useRef(null)

  useEffect(() => {
    ref.current.focus()
  }, [])
  
  const currentRanking = rankingLevels[rankingLevels.filter(i => Math.floor(geniusScore * i.minScoreMultiplier) <= currentScore).length - 1]
  const nextRanking = rankingLevels[rankingLevels.filter(i => Math.floor(geniusScore * i.minScoreMultiplier) <= currentScore).length]
  return (
    <div data-testid="menu-page" className="h-screen z-40 shadow w-full backdrop-blur-sm absolute flex flex-col items-center">
    <div tabIndex={-1}  ref={ref} onBlur={() => setTimeout(() => setShowMenuItem(null), 120)} data-testid="rankings-div" className="rounded mx-2 mt-24 p-4 pb-8 md:mx-24 lg:mx-60 shadow-md border flex flex-col  bg-white text-zinc-700 dark:bg-zinc-900 dark:text-zinc-300 dark:border dark:border-zinc-800">

      <div  className='flex flex-row justify-between '>
        <h2 className="font-bold text-2xl font-display">Rankings</h2>
        <button data-testid="menu-icon" className='hover:bg-gray-100 active:bg-gray-200 text-2xl m-2 w-10 h-10 rounded-full flex items-center justify-center' onClick={() => setShowMenuItem(null)}>
          <AiOutlineClose />
        </button>
      </div>
      <div data-testid="help-text" className='p-4 font-semithin text-lg'>   
        <p>Ranks are based on a percentage of possible points in a puzzle.</p>
      </div>
      <div>
        <div className='flex flex-row font-bold justify-between px-2'>
          <p className='ml-12 '>Rank</p>
          <p >Minimum score</p>
        </div>
        <div className=' '>
          {rankingLevels.map(i => (
            <div data-testid={`${i.name}-level`} key={i.name} className={`${i.minScoreMultiplier === currentRanking.minScoreMultiplier ? "bg-yellow-300 rounded-full font-bold" : i.minScoreMultiplier < currentRanking.minScoreMultiplier ? "text-gray-400" : ""} px-4 py-2 flex flex-row justify-between items-center`}>
              {i.minScoreMultiplier === currentRanking.minScoreMultiplier 
              ? <div className='m-2 flex '>
                  <p>{currentScore}</p>
                </div>
              : <div className={` w-2 h-2 m-2 mx-4 ${i.minScoreMultiplier < currentRanking.minScoreMultiplier ? "bg-yellow-300 rounded-full" : i.minScoreMultiplier < 1 ? "bg-gray-300 rounded-full" :  "bg-gray-300"}  flex items-center justify-center`} />}
              <div>
                <p>{i.name}</p>
                {i.minScoreMultiplier === currentRanking.minScoreMultiplier 
                && <div className='text-xs font-normal'>
                    <p>{Math.floor(geniusScore *  nextRanking.minScoreMultiplier) - currentScore} points to next rank, {geniusScore - currentScore} points to Genius</p>
                  </div>}
              </div>
              <div className=' flex flex-grow items-center px-4'>
                {i.name !== currentRanking.name && <hr className='hidden md:flex w-full'/>}
              </div>
              <p>{Math.floor(i.minScoreMultiplier * geniusScore)}</p>
            </div>
          )).reverse()}
        </div>
      </div>
    </div></div>
    
  )
}