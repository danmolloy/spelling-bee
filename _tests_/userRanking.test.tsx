import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom"
import UserRanking, { UserRankingProps, getRanking } from "../components/userRanking";
import { mockData } from "../_mocks_/gameData";
import { rankingLevels } from "../components/menu/rankings";

const mockProps: UserRankingProps = {
  answers: mockData.answers,
  userPoints: 20,
  isLoading: false
}

describe("UserRanking component", () => {
  beforeEach(() => {
    render(<UserRanking {...mockProps} />)
  })
  it("user-ranking", () => {
    const userRanking = screen.getByTestId("user-ranking")
    expect(userRanking).toBeInTheDocument()
  })
  it("ranking level name is in the document", () => {
    const userRanking = screen.getByTestId("user-ranking")
    expect(userRanking.textContent).toMatch(rankingLevels[getRanking(mockProps.userPoints, mockProps.answers)].name)
  })
  it("user points count is in the document", () => {
    const userRanking = screen.getByTestId("user-ranking")
    expect(userRanking.textContent).toMatch(String(mockProps.userPoints))
  })
})

describe("UserRanking component", () => {
  const mockProps: UserRankingProps = {
    answers: ["cat"],
    userPoints: 20,
    isLoading: false
  }
  beforeEach(() => {
    render(<UserRanking {...mockProps} />)
  })
  it("matches snapshot", () => {
    const userRanking = screen.getByTestId("user-ranking")
    expect(userRanking).toMatchSnapshot()
  })
})