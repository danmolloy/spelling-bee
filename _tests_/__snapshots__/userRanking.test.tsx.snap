// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserRanking component matches snapshot 1`] = `
<div
  class="flex flex-row w-full items-center justify-center"
  data-testid="user-ranking"
>
  <h3
    class="flex items-center justify-center p-1  w-32 font-semibold"
  >
    Genius
  </h3>
  <div
    class="w-full"
  >
    <div
      class="ranking-bar"
    >
      <div>
        <div
          class="past-rank-icon"
        />
      </div>
      <div>
        <div
          class="past-rank-icon"
        />
      </div>
      <div>
        <div
          class="past-rank-icon"
        />
      </div>
      <div>
        <div
          class="past-rank-icon"
        />
      </div>
      <div>
        <div
          class="past-rank-icon"
        />
      </div>
      <div>
        <div
          class="past-rank-icon"
        />
      </div>
      <div>
        <div
          class="past-rank-icon"
        />
      </div>
      <div>
        <div
          class="past-rank-icon"
        />
      </div>
      <div>
        <div
          class="current-rank-icon bounce"
        >
          <p
            class="font-thin text-sm "
          >
            20
          </p>
        </div>
      </div>
    </div>
    <div
      class="line"
    />
  </div>
</div>
`;
